### PR TITLE
network: more fd cleanup

### PR DESCRIPTION
**Pull request by florincoras**
- added shudown io handle api
- replaced io handle fd validations with isOpen
- changed proxy protocol to avoid use of ioctl
- more recv cleanup

Signed-off-by: Florin Coras <fcoras@cisco.com>

Risk Level: Medium (because of proxy protocol)
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by arthuryan-k** [comment]
LGTM, glad that proxy_protocol is being changed to remove use of ioctl - will update the fuzz target to reflect those changes once this is merged.

File: source/common/event/file_event_impl.cc
Comment (antoniovicente) [inline source/common/event/file_event_impl.cc:28]: It can also happen in the case of programmer error.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc
Comment (antoniovicente) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:278]: I'm at a loss here: why is it safe to skip these os_syscalls.ioctl calls that we used to have here?

It may make sense to move changes to this file to a different PR.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc
Comment (antoniovicente) [inline source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:279]: Seems like we have some test coverage gaps here, this throw doesn't seem to have an associated test.

File: test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc
Comment (mattklein123) [inline test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:363]: nit: I don't think this comment adds much value, same below.

Comment (florincoras) [reply test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:363]: Fixed and changed name of tests and comments.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc
Comment (florincoras) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:278]: ioctl was used to determine how much data is available for reading but, even ignoring the scenario where data is written after ioctl returns and before recv is called, the result is somewhat redundant because non-blocking fds return EAGAIN when there's no data to read. 

The change is to just call recv with buffer length set to max needed (or max buflen) and if EAGAIN return what we did when ioctl determined there was nothing to read, i.e., false. All other socket errors are treated as exceptions and whatever recv returned (needs to be positive) is what was successfully read. So, we don't really need to know how much is available for reading. 

Will move to separate PR to simplify this one.

File: source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc
Comment (florincoras) [reply source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc:279]: We're still throwing the same error on line 284 on recv errors (tests were changed to throw recv errors instead of ioctl)

File: source/common/event/file_event_impl.cc
Comment (florincoras) [reply source/common/event/file_event_impl.cc:28]: Is it okay if I just remove the comment?

File: source/common/event/file_event_impl.cc
Comment (mattklein123) [inline source/common/event/file_event_impl.cc:27]: I still think the comment is valuable, in some form, but will defer to @antoniovicente on that.

Comment (antoniovicente) [reply source/common/event/file_event_impl.cc:27]: Keeping the comment SGTM.

**Review by antoniovicente — DISMISSED** [review]
Thanks for the fd() cleanups.

File: source/common/event/file_event_impl.cc
Comment (florincoras) [reply source/common/event/file_event_impl.cc:29]: Done!

**Pull request merged**
